File: parser/shokparser/Parser.py
# Copyright (C) 2013 Michael Biggs.  See the COPYING file at the top-level
# directory of this distribution and at http://shok.io/code/copyright.html

import MakeRule
import Rule
import logging

class Parser(object):
  def __init__(self,rule,parent=None):
    self.rule = rule
    self.name = rule.name
    self.bad = self.rule.bad
    self.done = self.rule.done
    self.evil = self.rule.evil
    if parent:
      self.parent = parent
      self.top = parent.top
    else:
      self.parent = None
      self.top = self

  def __repr__(self):
    return '%s (%s)' % (self.name, self.rule.name)

def MakeParser(x, parent=None):
  if isinstance(x, Parser):
    return x
  elif isinstance(x, Rule.Rule):
    return x.MakeParser(parent)
  return MakeParser(MakeRule.MakeRule(x), parent)

File: p2/Parser.cpp
// Copyright (C) 2014 Michael Biggs.  See the COPYING file at the top-level
// directory of this distribution and at http://shok.io/code/copyright.html

#include "Parser.h"

#include "Shok.h"

#include <boost/bind.hpp>
#include <boost/spirit/include/phoenix_core.hpp>
#include <boost/spirit/include/phoenix_operator.hpp>
#include <boost/spirit/include/qi.hpp>
#include <boost/spirit/include/support_multi_pass.hpp>
namespace phoenix = boost::phoenix;
namespace spirit = boost::spirit;
namespace ascii = spirit::ascii;
namespace qi = spirit::qi;

#include <istream>
#include <ostream>
#include <string>
using std::endl;
using std::istream;
using std::ostream;
using std::string;

using namespace parser;

Parser::Parser(istream& input, ostream& output)
  : m_input(input),
    m_output(output) {
}

bool Parser::parse() {
  // iterate over stream input
  typedef std::istreambuf_iterator<char> base_iterator_type;
  base_iterator_type in_begin(m_input);

  // convert input iterator to forward iterator, usable by spirit parser
  typedef spirit::multi_pass<base_iterator_type> forward_iterator_type;
  forward_iterator_type fwd_begin = spirit::make_default_multi_pass(in_begin);
  forward_iterator_type fwd_end;

  ShokParser<forward_iterator_type> shok_(m_output);

  //BOOST_SPIRIT_DEBUG_NODE(shok_);

  bool r = qi::phrase_parse(
    fwd_begin, fwd_end,
    shok_,
    ascii::space
  );

  if (!r || fwd_begin != fwd_end) {
    return false;
  }

  return r;
}
